crates/bevy_reflect/src/serde/deserialize_reflect.rs

```diff
@@ -0,0 +1,92 @@
+use crate::{FromType, Reflect, TypeRegistry};
+use serde::{Deserialize, Deserializer};
+
+/// Trait used to provide finer control when deserializing a reflected type with one of
+/// the reflection deserializers.
+///
+/// This trait is the reflection equivalent of `serde`'s [`Deserialize`] trait.
+/// The main difference is that this trait provides access to the [`TypeRegistry`],
+/// which means that we can use the registry and all its stored type information
+/// to deserialize our type.
+///
+/// This can be useful when writing a custom reflection deserializer where we may
+/// want to handle parts of the deserialization process, but temporarily pass control
+/// to the standard reflection deserializer for other parts.
+///
+/// For the serializtion equivalent of this trait, see [`SerializeReflect`].
+///
+/// # Rationale
+///
+/// Without this trait and its associated [type data], such a deserializer would have to
+/// write out all of the deserialization logic itself, possibly including
+/// unnecessary code duplication and trivial implementations.
+///
+/// This is because a normal [`Deserialize`] implementation has no knowledge of the
+/// [`TypeRegistry`] and therefore cannot create a reflection-based deserializer for
+/// nested items.
+///
+/// # Implementors
+///
+/// In order for this to work with the reflection deserializers like [`TypedReflectDeserializer`]
+/// and [`UntypedReflectDeserializer`], implementors should be sure to register the
+/// [`ReflectDeserializeReflect`] type data.
+/// This can be done [via the registry] or by adding `#[reflect(DeserializeReflect)]` to
+/// the type definition.
+///
+/// Note that this trait has a blanket implementation for all types that implement
+/// [`Reflect`] and [`Deserialize`] which just calls the normal [`Deserialize`] implementation.
+///
+/// [`SerializeReflect`]: crate::serde::SerializeReflect
+/// [type data]: ReflectDeserializeReflect
+/// [`TypedReflectDeserializer`]: crate::serde::TypedReflectDeserializer
+/// [`UntypedReflectDeserializer`]: crate::serde::UntypedReflectDeserializer
+/// [via the registry]: TypeRegistry::register_type_data
+pub trait DeserializeReflect<'de>: Reflect + Sized {
+    fn deserialize<D>(deserializer: D, registry: &TypeRegistry) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>;
+}
+
+impl<'de, T: Reflect + Deserialize<'de>> DeserializeReflect<'de> for T {
```

Comment:
Not a dumb question! Yes, this will prevent custom impls, **but only where a `Deserialize` impl already exists**.

The reasoning is that if you can deserialize a value without the `TypeRegistry`, you shouldn't need the `DeserializeReflect` trait.

Of course, there could be users who want to have a separate reflection-based deserialization impl from their normal serde impl. But I don't forsee that being a common practice or at least one we may want to support, as it means we'd probably need to keep both `ReflectDeserialize` _and_ `ReflectDeserializeReflect`.

By using the blanket impl, we reduce the friction for when we do actually merge the two concepts (whether that happens in this PR or a future one).

I'm open to holding off on this until that happens though (assuming it doesn't happen in this PR).

Thoughts?

Reply:
I guess in that case I'd prefer to not adding the blanket implementation until it is actually needed. 
As you said, if the value is `Deserialize` we shouldn't need to implement `DeserializeReflect` :)